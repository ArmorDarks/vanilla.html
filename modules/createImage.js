const sharp = require('sharp')
const { mkdirSync, existsSync } = require('fs')
const { resolve, join, parse } = require('path')

const mock = (path) => {
  const image = sharp(path)
  image.done = async () => {
    return Promise.resolve(await image.metadata())
  }
  return image
}

const createImage = (
  src,
  {
    prefix = '',
    filename,
    postfix = '',
    outputDir,
    cwd = '',
    build,
    override
  }
) => {
  if (typeof src !== 'string') throw new Error(`[createImage] \`src\` should be a string, \`${typeof src}\` provided`)
  if (!build) return mock()

  let { dir, name, ext } = parse(src)

  if (outputDir) dir = outputDir

  const absDir = resolve(cwd, dir)
  const absSrc = resolve(cwd, src)

  if (!existsSync(absSrc)) throw new Error(`[createImage] src \`${absSrc}\` does not exist`)

  if (filename) name = filename

  const file = `${prefix}${name}${postfix}${ext === '.svg' ? '.png' : ext}`
  const output = join(dir, file)
  const absOutput = resolve(output)

  if (!override && existsSync(absOutput)) return mock(absOutput)
  if (!existsSync(absDir)) mkdirSync(absDir)

  const image = sharp(absSrc, { failOnError: true })

  const data = {
    src: output,
    prefix,
    filename,
    postfix,
    outputDir,
    cwd,
    build
  }

  image.done = async () => {
    let metadata

    try {
      metadata = await image.toFile(absOutput)
    } catch (error) {
      error.message = `[createImage]: ${error.message}`
      throw error
    }

    return Promise.resolve(Object.assign(data, metadata))
  }

  return image
}

module.exports = createImage
