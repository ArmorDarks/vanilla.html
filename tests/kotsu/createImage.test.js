/* eslint-env jest */

import createImage from '../../modules/createImage'
import { readFileSync } from 'fs'
// import rimraf from 'rimraf'
import { resolve, join } from 'path'
// import { promisify } from 'util'

const filename = 'image'
const ext = 'png'
const file = `${filename}.${ext}`
const src = join('tests/kotsu/fixtures', file)
const outputDir = 'temp/images'

describe('createImage function', () => {
  // beforeAll(() => promisify(rimraf)(resolve(outputDir)))

  it('should resize image', async () => {
    const postfix = '@resize'
    const output = join(outputDir, `${filename}${postfix}.${ext}`)

    await createImage(src, {
      outputDir,
      postfix,
      build: true
    })
      .resize(200)
      .done()

    expect(JSON.stringify(readFileSync(output))).toMatchSnapshot()
  })

  it('should have proper stats', async () => {
    await createImage(src, {
      outputDir,
      postfix: '@metadata',
      build: true
    })
      .resize(200)
      .done()
      .then((metadata) => expect(metadata).toMatchSnapshot())
  })
})
