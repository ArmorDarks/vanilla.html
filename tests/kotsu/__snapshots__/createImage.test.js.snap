// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createImage function should have proper stats 1`] = `
Object {
  "build": true,
  "channels": 3,
  "cwd": "",
  "filename": undefined,
  "format": "png",
  "height": 200,
  "outputDir": "temp/images",
  "postfix": "@metadata",
  "prefix": "",
  "premultiplied": false,
  "size": 447,
  "src": "temp\\\\images\\\\image@metadata.png",
  "width": 200,
}
`;

exports[`createImage function should resize image 1`] = `"{\\"type\\":\\"Buffer\\",\\"data\\":[137,80,78,71,13,10,26,10,0,0,0,13,73,72,68,82,0,0,0,200,0,0,0,200,8,2,0,0,0,34,58,57,201,0,0,0,9,112,72,89,115,0,0,11,18,0,0,11,18,1,210,221,126,252,0,0,1,113,73,68,65,84,120,218,237,210,49,13,0,0,12,195,176,242,39,221,162,152,180,195,134,16,37,133,3,145,0,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,48,22,24,11,99,97,44,48,22,198,194,88,96,44,140,133,177,192,88,24,11,99,129,177,48,22,198,2,99,97,44,140,5,198,194,88,24,11,140,133,177,48,22,24,139,111,6,194,202,6,158,154,175,26,42,0,0,0,0,73,69,78,68,174,66,96,130]}"`;
